Add contributing guidelines

TortillasAlfred opened the pull request:
Add CONTRIBUTING.md file to provide basic guidelines on how to contribute to Poutyne.

freud14 commented:
The license should be GPLv3.

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+## We Use [Github Flow](https://guides.github.com/introduction/flow/index.html), So All Code Changes Happen Through Pull Requests
+Pull requests are the best way to propose changes to the codebase. We actively welcome your pull requests:
+
+1. Fork the repo and create your branch from `master`.
```
Should we propose a branch name pattern ?
Something like: `fixing_bug-... `

TortillasAlfred replied:
Since we suggest to fork the repository, the contributor never creates a branch on the Poutyne git, only on his personal workspace. When submitting the PR, he then proposes changes coming from his own fork and therefore branch naming is not really an issue here.

Correct me if i'm wrong @freud14 .

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+Pull requests are the best way to propose changes to the codebase. We actively welcome your pull requests:
+
+1. Fork the repo and create your branch from `master`.
+2. If you've added code that should be tested, add tests.
```
If you've added new code, then it **must** be tested. 
­

TortillasAlfred replied:
```suggestion
2. If you've added code that should be tested, you **must** ensure it is properly tested.
```

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+1. Fork the repo and create your branch from `master`.
+2. If you've added code that should be tested, add tests.
+3. If you've changed APIs, update the documentation.
+4. Ensure the test suite passes.
```
Should not we have a travis test suite build up for that ? (removing the concept to our automation system)

TortillasAlfred replied:
We already have a Travis set up. It's more of a "please ensure the Travis tests did not fail".

TortillasAlfred replied:
```suggestion
4. Ensure the Travis test suite passes.
```

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+2. If you've added code that should be tested, add tests.
+3. If you've changed APIs, update the documentation.
+4. Ensure the test suite passes.
+5. Make sure your code lints.
```
Should not we use a linter in the .yml of travis ? (easier and more consistent with our style)

TortillasAlfred replied:
I'm no expert with Travis. Can we automatically run the linting and/or the formatting with Travis. 

Let's say we add the linting, how does the contributor see the results ?

freud14 replied:
We probably can run the linting in Travis but I don't know if there is a simple way to make it fail the tests if linting does not pass. Also, if we do that, the pylint version should be fixed to some version because each version tends to add warnings that were not there previously.

With yapf, is it possible to just test the formatting and not modify the files? If so, it can be a good idea to add it to the Travis script.

Anyway, I don't think that these modifications fall in the scope of this pull request.

TortillasAlfred replied:
Agreed, the Travis integration is interesting but out of scope here.

davebulaval replied:
I'll take a look for the linter integration also, just to be sure we don't miss any bad syntax.

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+3. If you've changed APIs, update the documentation.
+4. Ensure the test suite passes.
+5. Make sure your code lints.
+6. Issue that pull request!
```
Is it necessary ?

TortillasAlfred replied:
Issue that PR => submit it. I'll change the name.

TortillasAlfred replied:
```suggestion
6. Submit that pull request!
```

davebulaval replied:
Oh I thought it was the concept of issue.

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+## Report bugs using Github's [issues](https://github.com/GRAAL-Research/poutyne/issues)
+We use GitHub issues to track public bugs. Report a bug by [opening a new issue](); it's that easy!
+
+## Write bug reports with detail, background, and sample code
```
*###

TortillasAlfred replied:
Why the third '#' ? I feel like it is okay for this title to be at same emphasis level as 'Use a Consistent Coding Style' and family.

davebulaval replied:
The third will put the concept into the title report bugs...

TortillasAlfred replied:
Have a look at 649a7bf483f1742f57c0ebf446ce449b4c88d159, I think it clears the confusion there.

davebulaval replied:
Better

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+
+**Great Bug Reports** tend to have:
+
+- A quick summary and/or background
```
and clear

TortillasAlfred replied:
```suggestion
- A quick and clear summary and/or background
```

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+
+- A quick summary and/or background
+- Steps to reproduce
+  - Be specific!
```
and clear

TortillasAlfred replied:
```suggestion
  - Be specific and clear!
```

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+- What you expected would happen
+- What actually happens
+- Notes (possibly including why you think this might be happening, or stuff you tried that didn't work)
+
```
Feel free ton include print screen also for more clarity.

TortillasAlfred replied:
```suggestion
Feel free to include any print screen or other file you feel may further clarify your point.
```

davebulaval commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,86 @@
+```
+pylint poutyne
+```
+
````
As stated before, should not this step been include into our build phase ?

davebulaval commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,86 @@
+```
+pytest tests
+```
+
````
Suggestion from [AllenNLP How To](https://github.com/allenai/allennlp/blob/master/CONTRIBUTING.md)
Do you have a suggestion for an enhancement?

We use GitHub issues to track enhancement requests. Before you create an enhancement request:

    Make sure you have a clear idea of the enhancement you would like. If you have a vague idea, consider discussing it first on the users list.

    Check the documentation to make sure your feature does not already exist.

    Do a quick search to see whether your enhancement has already been suggested.

When creating your enhancement request, please:

    Provide a clear title and description.

    Explain why the enhancement would be useful. It may be helpful to highlight the feature in other libraries.

    Include code examples to demonstrate how the enhancement would be used.

davebulaval commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+- What actually happens
+- Notes (possibly including why you think this might be happening, or stuff you tried that didn't work)
+
+## Use a Consistent Coding Style
```
```suggestion
## Do you have a suggestion for an enhancement?

We use GitHub issues to track enhancement requests.  Before you create an enhancement request:

* Make sure you have a clear idea of the enhancement you would like.  If you have a vague idea, consider discussing
it first on the users list.

* Check the documentation to make sure your feature does not already exist.

* Do [a quick search](https://github.com/allenai/allennlp/issues) to see whether your enhancement has already been suggested.

When creating your enhancement request, please:

* Provide a clear title and description.

* Explain why the enhancement would be useful.  It may be helpful to highlight the feature in other libraries.

* Include code examples to demonstrate how the enhancement would be used.

## Use a Consistent Coding Style
```

davebulaval replied:
From [AllenNLP](https://github.com/allenai/allennlp/blob/master/CONTRIBUTING.md)

TortillasAlfred commented:
Updated to the correct license and fixed most comments.

I am not quite sure about what to do with the Travis test suite regarding the inclusion of linting/formatting. What are your ideas guys ?

freud14 commented:
As I said in the thread above, I don't think that these modifications fall in the scope of this pull request.

davebulaval commented:
For myself, I'm ok with the change.
